src/libraries/System.Linq/src/System/Linq/Shuffle.SpeedOpt.cs

```diff
@@ -0,0 +1,272 @@
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+
+namespace System.Linq
+{
+    public static partial class Enumerable
+    {
+        private sealed partial class ShuffleIterator<TSource>
+        {
+            public override TSource[] ToArray()
+            {
+                TSource[] array = _source.ToArray();
+                Random.Shared.Shuffle(array);
+                return array;
+            }
+
+            public override List<TSource> ToList()
+            {
+                List<TSource> list = _source.ToList();
+                Random.Shared.Shuffle(CollectionsMarshal.AsSpan(list));
+                return list;
+            }
+
+            public override int GetCount(bool onlyIfCheap) =>
+                !onlyIfCheap ? _source.Count() :
+                TryGetNonEnumeratedCount(_source, out int count) ? count :
+                -1;
+
+            public override TSource? TryGetFirst(out bool found) =>
+                TryGetElementAt(0, out found);
+
+            public override TSource? TryGetLast(out bool found) =>
+                TryGetElementAt(0, out found);
+
+            public override TSource? TryGetElementAt(int index, out bool found)
+            {
+                if (_source is Iterator<TSource> iterator &&
+                    iterator.GetCount(onlyIfCheap: true) is int iteratorCount &&
+                    iteratorCount >= 0)
+                {
+                    if ((uint)index < (uint)iteratorCount)
+                    {
+                        return iterator.TryGetElementAt(Random.Shared.Next(0, iteratorCount), out found);
+                    }
+                }
+                else if (_source is IList<TSource> list)
+                {
+                    int listCount = list.Count;
+                    if ((uint)index < (uint)listCount)
+                    {
+                        found = true;
+                        return list[Random.Shared.Next(0, listCount)];
+                    }
+                }
+                else if (index >= 0)
+                {
+                    TSource[] array = _source.ToArray();
+                    if (index < array.Length)
+                    {
+                        found = true;
+                        return array[Random.Shared.Next(0, array.Length)];
+                    }
+                }
+
+                found = false;
+                return default;
+            }
+
+            public override Iterator<TSource>? Take(int count)
+            {
+                // If the source is known to have fewer elements than count, we're best off just using the default implementation.
+                if (_source.TryGetNonEnumeratedCount(out int sourceCount) && sourceCount <= count)
+                {
+                    return base.Take(count);
+                }
+
+                // Otherwise, we either don't know how many elements are in the source, or we know it's more than count.
```

Comment:
My understanding is that builds of System.Linq that have this optimization enabled will behave differently when calling `Shuffle().Take()` on infinite enumerables. Are we ok with this?

Reply:
Correspondingly `infiniteEnumerable.Shuffle().Take(1).Count()` completes whereas `await infiniteEnumerable.AsAsyncEnumerable().Shuffle().Take(1).CountAsync()` diverges.

Reply:
We've now shifted away from two different builds and instead just have a single build with a feature switch, so it's really a question of whether the feature switch impacts behavior, and that's much more palatable, since that's what feature switches are for :)

This whole file doesn't come into play if the size-optimized feature switch is set, so I'm not sure there's much we can do about it, unless you're suggesting blocking enumerables > 2^31 elements from working even when we can support them?